src/FillArrays.jl

```diff
@@ -198,7 +198,59 @@ rank(F::Zeros) = 0
 rank(F::Ones) = 1
 
 
-const Eye{T, Axes} = Diagonal{T, Ones{T,1,Tuple{Axes}}}
+struct RectDiagonal{T,V<:AbstractVector{T},Axes<:Tuple{Vararg{AbstractUnitRange,2}}} <: AbstractMatrix{T}
+    diag::V
+    axes::Axes
+
+    @inline function RectDiagonal{T,V}(A::V, axes::Axes) where {T,V<:AbstractVector{T},Axes<:Tuple{Vararg{AbstractUnitRange,2}}}
+        @assert !Base.has_offset_axes(A)
+        @assert any(length(ax) == length(A) for ax in axes)
+        rd = new{T,V,Axes}(A, axes)
+        @assert !Base.has_offset_axes(rd)
+        return rd
+    end
+end
+
+@inline RectDiagonal{T,V}(A::V, sz::Tuple{Vararg{Integer, 2}}) where {T,V} = RectDiagonal{T,V}(A, Base.OneTo.(sz))
+@inline RectDiagonal{T,V}(A::V, axes::Vararg{Any, 2}) where {T,V} = RectDiagonal{T,V}(A, axes)
+@inline RectDiagonal{T,V}(A::V, sz::Vararg{Integer, 2}) where {T,V} = RectDiagonal{T,V}(A, sz)
+@inline RectDiagonal{T,V}(A::V) where {T,V} = RectDiagonal{T,V}(A, (axes(A, 1), axes(A, 1)))
+@inline RectDiagonal{T}(A::V, args...) where {T,V} = RectDiagonal{T,V}(A, args...)
+@inline RectDiagonal(A::V, args...) where {V} = RectDiagonal{eltype(V),V}(A, args...)
+
+axes(rd::RectDiagonal) = rd.axes
+size(rd::RectDiagonal) = length.(rd.axes)
+
+@inline function getindex(rd::RectDiagonal{T}, i::Integer, j::Integer) where T
+    @boundscheck checkbounds(rd, i, j)
+    if i == j
+        @inbounds r = rd.diag[i]
+    else
+        r = zero(T)
+    end
+    return r
+end
+
+function setindex!(rd::RectDiagonal, v, i::Integer, j::Integer)
+    @boundscheck checkbounds(rd, i, j)
+    if i == j
+        @inbounds rd.diag[i] = v
+    elseif !iszero(v)
+        throw(ArgumentError("cannot set off-diagonal entry ($i, $j) to a nonzero value ($v)"))
+    end
+    return v
+end
+
+diag(rd::RectDiagonal) = rd.diag
+
+for f in (:triu, :triu!, :tril, :tril!)
+    @eval ($f)(M::RectDiagonal) = M
+end
+
+
+const RectOrDiagonal{T,V,Axes} = Union{RectDiagonal{T,V,Axes}, Diagonal{T,V}}
+const SquareEye{T,Axes} = Diagonal{T,Ones{T,1,Tuple{Axes}}}
+const Eye{T,Axes} = RectOrDiagonal{T,Ones{T,1,Tuple{Axes}}}
```

Comment:
Looks like the change in type broke a test in InfiniteArrays (see [the JuliaCIBot results](https://juliarun-ci.s3.amazonaws.com/d0b0118e8c905b30b1d1ed15ef578d04ed55bbe5/InfiniteArrays_with_pull_request_of_FillArrays_on_julia_0_7.log)).

Reply:
I have a fix for it, I'll submit a PR to InfiniteArrays.

Reply:
Thanks!